Add EIP-7251 spec: Increase MAX_EFFECTIVE_BALANCE

dapplion opened the pull request:
Consensus implementation for [EIP-7251:  Increase MAX_EFFECTIVE_BALANCE](https://eips.ethereum.org/EIPS/eip-7251)

Extensive discussion and rationale can be found in this PR:
- https://github.com/michaelneuder/consensus-specs/pull/3

Accompanying documentation:
- Original ethresearch proposal: https://ethresear.ch/t/increase-the-max-effective-balance-a-modest-proposal/15801
- Security considerations: https://notes.ethereum.org/@fradamt/meb-increase-security
- FAQ: https://notes.ethereum.org/@mikeneuder/eip-7251-faq

ensi321 commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,827 @@
+    validator.withdrawable_epoch = max(validator.withdrawable_epoch, Epoch(epoch + EPOCHS_PER_SLASHINGS_VECTOR))
+    state.slashings[epoch % EPOCHS_PER_SLASHINGS_VECTOR] += validator.effective_balance
+    slashing_penalty = validator.effective_balance // MIN_SLASHING_PENALTY_QUOTIENT_EIP7251  # [Modified in EIP7251]
+    decrease_balance(state, slashed_index, slashing_penalty)
```
It looks like the whistleblower reward part is missing/removed. Is that intentional?

mkalinin replied:
A rough consensus is to set `MIN_SLASHING_PENALTY_QUOTIENT_EIP7251` and whistleblower reward to the same value to avoid minting ETH. I am planning to propose this change in the next few days

prestonvanloon commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,827 @@
+            withdrawal_index += WithdrawalIndex(1)
+            consumed += 1
+
+    state.pending_partial_withdrawals = state.pending_partial_withdrawals[consumed:] 
```
This modifies the state in a "getter" function. That update should happen elsewhere.

tuyennhv commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,887 @@
+    # only exit validator if it has no pending withdrawals in the queue
+    if is_full_exit_request and pending_balance_to_withdraw == 0:
+        initiate_validator_exit(state, index)
+    elif state.balances[index] > MIN_ACTIVATION_BALANCE + pending_balance_to_withdraw:
```
```suggestion
    elif state.balances[index] >= MIN_ACTIVATION_BALANCE + pending_balance_to_withdraw:
```

mkalinin replied:
In the equality case there is nothing left to withdraw so the withdrawal shouldn’t be created

mkalinin commented:
#### `MIN_SLASHING_PENALTY_QUOTIENT_EIP7251 = 65536` rationale

With this value for the quotient the initial slashing penalty will be roughly equal to 6 days of base_reward (with 32M ETH of total active balance), i.e. it is roughly 6 days worth of attestation rewards. In absolute values it will be `0.03125` ETH and `0.00050` ETH for 2048 and 32 ETH validators respectively. The main objective of this reduction is to alleviate additional risk that consolidated validators will have comparing to those that are not consolidated.

Whistleblower reward is also updated: `WHISTLEBLOWER_REWARD_QUOTIENT_EIP7251 = 65536` which is 6 days worth of attestation rewards. This should still provide enough incentives to submit slashings and being capped at the `MIN_SLASHING_PENALTY_QUOTIENT_EIP7251` will prevent new ETH minting when slashing is processed.

The script to compute a number of base rewards equal to a slashing penalty:
```python
import math

BASE_REWARD_FACTOR = 64
EFFECTIVE_BALANCE_INCREMENT = 1_000_000_000
TOTAL_ACTIVE_BALANCE = 32_000_000_000 * 1_000_000_000  # 32M ETH

MIN_SLASHING_PENALTY_QUOTIENT_EIP7251 = 2**16


def get_base_reward_per_increment():
    return EFFECTIVE_BALANCE_INCREMENT * BASE_REWARD_FACTOR // math.floor(TOTAL_ACTIVE_BALANCE ** 0.5)


def get_base_reward(effective_balance: int):
    increments = effective_balance // EFFECTIVE_BALANCE_INCREMENT
    return increments * get_base_reward_per_increment()


def get_base_rewards_in_slashing_penalty(effective_balance: int):
		base_reward = get_base_reward(effective_balance)
		slashing_penalty = effective_balance // MIN_SLASHING_PENALTY_QUOTIENT_EIP7251
		return slashing_penalty // base_reward
```

dapplion commented on specs/_features/eip7251/beacon-chain.md:
````diff
@@ -0,0 +1,896 @@
+##### Updated `get_expected_withdrawals`
+
+```python
+def get_expected_withdrawals(state: BeaconState) -> Tuple[Sequence[Withdrawal], uint64]:
````
what do you mean with "tag"?

ensi321 replied:
> what do you mean with "tag"?

I mean like adding `# [Modified in EIP7251]` after the function signature to signal the return type has changed

dapplion commented:
@potuz I've added commit https://github.com/ethereum/consensus-specs/pull/3618/commits/ebdb513ec8789c0b3d5a4a8289e2f62814147a93 with `queue_excess_active_balance` on the fork transition. It queues all excess balance from early compounding credential adopters

@ethDreamer I've added https://github.com/ethereum/consensus-specs/pull/3618/commits/7ec93c565554ffcc176f6945ba7b4ce456195649 to switch the target consolidator into compounding credentials if the consolidation is sucessful

dapplion commented:
This PR intentionally does not modify `EJECTION_BALANCE`. I will share a write-up shortly justifying why 🙏

dapplion commented:
### `is_aggregator` changes

I have removed the `is_aggregator` changes. While balance-based aggregator selection is necessary for safety, it is only applicable for high regimes of consolidation. The current spec includes consolidation churn, which guarantees a maximum (slow) rate of adoption. We should leave `is_aggregator` changes out now to reduce complexity later and include them in the subsequent fork if the stakers choose to consolidation into very concentrated validators. Will publish a doc soon with full rationale and numbers to back it up

### No p2p topic

This PR is missing a p2p doc on purpose. I believe we don't need a p2p topic to distribute consolidation objects:

Ethereum today relies on HTTP connections for two key actions:
- most (if not all) syncings from a blank DB (checkpoint sync)
- proposing +90% of blocks (with MEV-boost)

I don't think it would be a big ask to accept the distribution of consolidations via an off-protocol solution. Most big players can just self-include on their own proposals. There's no incentive to consolidate before or after than other actors. As long the network eventually consolidates. If the off-protocol message distribution service fails that's okay there's no loss of funds just longer inclusion times. Specially because consolidations are a temporary feature that's only useful during a transition. New stake can just deposit consolidated. Removing an HTTP API in 2 years is much simpler than removing p2p code.

hwwhww commented:
IIRC @fradamt has a branch with new EIP-7251 test cases: https://github.com/michaelneuder/consensus-specs/pull/20

Can we add them to this PR now?

fradamt commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,907 @@
+                              withdrawal_credentials: Bytes32,
+                              amount: uint64) -> None:
+    index = get_index_for_new_validator(state)
+    validator = get_validator_from_deposit(pubkey, withdrawal_credentials, amount)
```
```suggestion
    validator = get_validator_from_deposit(pubkey, withdrawal_credentials)
```

fradamt commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,907 @@
+    set_or_append_list(state.previous_epoch_participation, index, ParticipationFlags(0b0000_0000))
+    set_or_append_list(state.current_epoch_participation, index, ParticipationFlags(0b0000_0000))
+    set_or_append_list(state.inactivity_scores, index, uint64(0))
+    state.pending_balance_deposits.append(PendingBalanceDeposit(index, amount))  # [New in EIP7251]
```
```suggestion
    state.pending_balance_deposits.append(PendingBalanceDeposit(index=index, amount=amount))  # [New in EIP7251]
```

fradamt commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,907 @@
+    else:
+        # Increase balance by deposit amount
+        index = ValidatorIndex(validator_pubkeys.index(pubkey))
+        state.pending_balance_deposits.append(PendingBalanceDeposit(index, amount))  # [Modified in EIP-7251]
```
```suggestion
        state.pending_balance_deposits.append(PendingBalanceDeposit(index=index, amount=amount))  # [Modified in EIP-7251]
```

rolfyone commented:
How does the interplay with other withdrawal types go?
Concrete example,
I'm a genesis validator (not, but as an example) and i've accrued 6 eth of balance since genesis, because i've never switched to `0x01` withdrawal key..
After electra at some point, I finally get around to issuing a bls change, so now Im a validator with 38 odd eth and a 0x01 creds... 
Does that mean that i get automatically switched to consolidated due to my more than one eth extra? Or would i have to specifically do a deposit?

dapplion commented:
> Does that mean that i get automatically switched to consolidated due to my more than one eth extra? Or would i have to specifically do a deposit?

No validator is automatically changed into consolidating credentials. The current paths to switch are:
- (implemented): through a consolidation operation, change target validator to 0x02 credentials
- (pending): through a top-up deposit that includes 0x02 credentials

avsetsin commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,907 @@
+    """
+    Check if ``validator`` is partially withdrawable.
+    """
+    return (
+        has_execution_withdrawal_credential(validator)  # [Modified in EIP7251]
+        and get_validator_excess_balance(validator, balance) > 0
+    )
```
In capella spec, there is a check in the `is_partially_withdrawable_validator` method:

```python
has_max_effective_balance = validator.effective_balance == MAX_EFFECTIVE_BALANCE
```

It is needed so that a validator with effective balance < `MAX_EFFECTIVE_BALANCE` can reach the balance = `MAX_EFFECTIVE_BALANCE + 0.25 eth` without being partially withdrawn. E.g. a validator with effective balance = 31 eth must have balance more than 32.25 eth to trigger updating his effective balance to 32 eth.


```suggestion
    if has_eth1_withdrawal_credential(validator) and validator.effective_balance == MIN_ACTIVATION_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    if has_compounding_withdrawal_credential(validator) and validator.effective_balance == MAX_EFFECTIVE_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    return False
```

mkalinin replied:
Good catch! I think we should refactor the spec a bit, and make the code of the sweep similar to Cancun, something like in this commit: https://github.com/mkalinin/eth2.0-specs/commit/4c7a65ac409ba653e531142f99558972483b84d2

dapplion commented:
@hwwhww can we merge this as-is so contributors can PR fixes on `ethereum/consensus-specs` instead of my random fork?

hwwhww reviewed (commented):
@dapplion 

I haven't reviewed it properly to merge it. Can @mkalinin @fradamt give some PR approvals here?

mkalinin reviewed (approved):
Outstanding work everyone! I think it’s a time to merge this PR and continue polishing and fixing the logic in separate PRs as it becomes inconvenient to introduce new changes to the feature

fradamt commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -429,6 +440,29 @@ def set_compounding_withdrawal_credentials(state: BeaconState, index: ValidatorI
+    balance = state.balances[index]
+    if balance > MAX_EFFECTIVE_BALANCE:
+        excess_balance = balance - MAX_EFFECTIVE_BALANCE
+        state.balances[index] = balance - excess_balance
```
feels clearer to just say:
```
state.balances[index] = MAX_EFFECTIVE_BALANCE
```

hwwhww replied:
```suggestion
        state.balances[index] = MAX_EFFECTIVE_BALANCE
```

It feels confusing to use both `MAX_EFFECTIVE_BALANCE` and `MAX_EFFECTIVE_BALANCE_EIP7251` in the same file. What do you think about adding a duplicate `MAX_EFFECTIVE_BALANCE_PRE_EIP7251 := 32`?

mkalinin replied:
I agree with that, maybe `MAX_EFFECTIVE_BALANCE_PHASE0`?

hwwhww replied:
oh yeah using `MIN_ACTIVATION_BALANCE` looks better to me 👍

mkalinin replied:
So you suggest the opposite, to replace `MAX_EFFECTIVE_BALANCE` with `MIN_ACTIVATION_BALANCE` everywhere in the spec? I am fine with either, as long as we use the same param name in all places

fradamt replied:
If the two options are replacing `MIN_ACTIVATION_BALANCE` with `MAX_EFFECTIVE_BALANCE_PHASE0` or replacing `MAX_EFFECTIVE_BALANCE` with `MIN_ACTIVATION_BALANCE`, I would prefer the latter, because `MIN_ACTIVATION_BALANCE` accurately represents what this parameter is in the current system

mkalinin replied:
> because `MIN_ACTIVATION_BALANCE` accurately represents what this parameter is in the current system

Not in every case, it is also a max effective balance for 0x01 creds. And when we want to get MAXEB of a particular validator it makes more sense to use `MAX_EFFECTIVE_BALANCE_PHASE) and *_EIP7251`. But again, I am fine with either, just don’t want to have `MAX_EB`, `MIN_AB` and `MAX_EB_EIP7251` in the spec as this is really confusing.

dapplion replied:
For bumped constants in the past we have used: 

```
INACTIVITY_PENALTY_QUOTIENT
INACTIVITY_PENALTY_QUOTIENT_ALTAIR
INACTIVITY_PENALTY_QUOTIENT_BELLATRIX
```

hwwhww reviewed (approved):
I haven't reviewed it carefully, but I am going to merge it to accelerate the later iterations.
